### PR TITLE
Zookeeper

### PR DESCRIPTION
**Pull request by vihang**

### CODE REVIEW

**Comment by mahendra** [comment]
@vihang a bit confused about this pull request. what is this for? have you made any changes? can't see any commits...

**Comment by vihang** [comment]
My bad. I made some changes but too soon to make a request. I will wait till this gets merged with Mumrahs Master.

**Comment by mahendra** [comment]
okay. do let me know what the changes are.

File: kafka/client.py [hunk 1]
```diff
@@ -160,8 +168,15 @@ def _send_broker_aware_request(self, payloads, encoder_fn, decoder_fn):
                                  correlation_id=requestId, payloads=payloads)
 
             # Send the request, recv the response
-            conn.send(requestId, request)
-            response = conn.recv(requestId)
+            try:
+                conn.send(requestId, request)
+                response = conn.recv(requestId)
+            except (socket.error, KafkaConnectionError) as exp:
```
Comment (timcherry) [inline kafka/client.py:178]: KafkaConnectionError is not imported

File: kafka/zookeeper.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+        # Start the producer instance
+        self.client = get_client(self.zkclient, chroot=chroot)
+        self.producer = self.producer_kls(self.client, topic, **kwargs)
+
+        # Stop Zookeeper
+        self.zkclient.stop()
+        self.zkclient.close()
+        self.zkclient = None
+
+    def stop():
```
Comment (timcherry) [inline kafka/zookeeper.py]: I think you forgot 'self' in the function prototype.

File: kafka/client.py (hunk 1 above)
Comment (mahendra) [inline kafka/client.py:178]: added

File: kafka/zookeeper.py (hunk 2 above)
Comment (mahendra) [inline kafka/zookeeper.py]: Good catch. fixed.

On Tue, Aug 13, 2013 at 12:05 AM, timcherry notifications@github.comwrote:

> In kafka/zookeeper.py:
> 
> > -        if self.producer_kls is None:
> > -            raise NotImplemented("Producer class needs to be mentioned")
> >   +
> > -        self.zkclient = KazooClient(hosts=hosts)
> > -        self.zkclient.start()
> >   +
> > -        # Start the producer instance
> > -        self.client = get_client(self.zkclient, chroot=chroot)
> > -        self.producer = self.producer_kls(self.client, topic, **kwargs)
> >   +
> > -        # Stop Zookeeper
> > -        self.zkclient.stop()
> > -        self.zkclient.close()
> > -        self.zkclient = None
> >   +
> > -    def stop():
> 
> I think you forgot 'self' in the function prototype.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mahendra/kafka-python/pull/1/files#r5717491
> .

## 

Mahendra

http://twitter.com/mahendra

**Comment by timcherry** [comment]
@mahendra  I have been looking at the 0.8 C(https://github.com/edenhill/librdkafka) and Java(https://git-wip-us.apache.org/repos/asf/kafka.git) clients for Kafka. In the C client there is never any connection made to the Zookeeper server. However, in the Java client a connection is opened to a zookeeper host, which requests meta data mapping (topic, partition) -> broker... Seeing as how kafka brokers are able to answer the same meta data request, what value does it add to have the clients communicate with zookeeper?

**Comment by timcherry** [comment]
Also, I think it might be useful to allow the ZSimpleProducer, ZSimpleConsumer, KafkaClient objects to accept a list of zookeeper/kafka hosts. And then make the meta data request to a random host in this list...This would help in load balancing. The Java client explains more about in this page: 

https://cwiki.apache.org/confluence/display/KAFKA/0.8.0+Producer+Example.

I have a patch for this if you allow me to write.

**Comment by mahendra** [comment]
@timcherry From a producer's perspective, we are using Zookeeper only to discover one active broker. After that the broker will talk to one broker, get the metadata and send messages to the appropriate broker (based on partitioning scheme used). Thus, Zookeeper support is used only for compatibility sake. Also, it helps admins that the clients talk to a Zookeeper node always. That way, the brokers can be moved around behind the scenes.

From a consumer's perspective, things are pretty much the same, but also Zookeeper provides a way for balancing partitions among different consumers.

**Comment by mahendra** [comment]
@timcherry ZSimpleProducer, ZSimpleConsumer do accept a list of hosts. If you see the argument in the code, it is not <tt>host</tt>, but <tt>hosts</tt>

Have a look at the examples. The hosts is a comma separated list of Zookeeper hosts.

Why a comma separated list? instead of <tt>[ 'host1', 'host2' ]</tt>
- Because the underlying <tt>kazoo</tt> library uses it in that format
- Also, it makes it easier to read/write the hosts configuration in a config file (text) and handle it. Makes an admin's life easier.

Regarding <tt>KafkaClient</tt>, as I mentioned in my previous comment - mentioning any one broker is enough. The library will query the metadata and do load-balancing accordingly.

**Pull request closed without merging**